### PR TITLE
strongly type `ActivityPeriod`

### PR DESCRIPTION
**Pull request by eladyn**
From my observations, `ActivityPeriod` **always** comes in one of the following three compositions:

- `decade`
- `start_year` (open timespan)
- `start_year` + `end_year` (closed timespan)

Thus, it probably makes sense to strongly type those variants into `ActivityPeriod`.

I also added two `debug_assert!`s that make sure that the invariants that I assume are always held up. Alternatively, I could also turn them into normal `assert!`s or handle-able exceptions through `TryFrom`. I went for the current variants, since I think it makes most sense, but I'm open to different perspectives!

I've verified my assumptions with the help of the below program for probably around 1000 different artists, so it should be fine? Some example outputs:

- for `spotify:artist:5WUlDfRSoLAfcVSX1WnrxN`: `known good artists: 470 (decades: 203, closed: 4, open: 31)`
- for `spotify:artist:3fMbdgg4jU18AjLCKBhRSm`: `known good artists: 333 (decades: 966, closed: 112, open: 216)`

<details>

```rust
use std::{
    collections::{HashSet, VecDeque},
    env,
    io::Write,
    process::exit,
};

use librespot_core::{spotify_id::SpotifyItemType, Session, SessionConfig, SpotifyId};
use librespot_discovery::Credentials;
use librespot_metadata::{artist::ActivityPeriod, Artist, Metadata};

#[tokio::main]
async fn main() {
    env_logger::init();
    let session_config = SessionConfig::default();

    let args: Vec<_> = env::args().collect();
    if args.len() != 4 && args.len() != 5 {
        eprintln!("Usage: {} USERNAME PASSWORD ARTIST_URI", args[0]);
        return;
    }
    let credentials = Credentials::with_password(&args[1], &args[2]);

    let uri = SpotifyId::from_uri(&args[3]).unwrap_or_else(|_| {
        eprintln!(
            "URI should be a URI such as: \
                \"spotify:artist:5WUlDfRSoLAfcVSX1WnrxN\""
        );
        exit(1);
    });

    if uri.item_type != SpotifyItemType::Artist {
        eprintln!("URI must be of item type artist!");
        exit(1);
    }

    let session = Session::new(session_config, None);
    if let Err(e) = session.connect(credentials, false).await {
        eprintln!("Error connecting: {}", e);
        exit(1);
    }

    let mut decades = 0;
    let mut closed_timespans = 0;
    let mut open_timespans = 0;
    let mut known_artist_ids: HashSet<SpotifyId> = HashSet::new();
    let mut new_artist_ids: VecDeque<SpotifyId> = VecDeque::new();
    new_artist_ids.push_back(uri);
    while let Some(id) = new_artist_ids.pop_front() {
        known_artist_ids.insert(id);
        let artist = Artist::get(&session, &id).await.unwrap();
        for period in artist.activity_periods.0 {
            match period {
                ActivityPeriod::Decade(_) => decades += 1,
                ActivityPeriod::Timespan { end_year, .. } => match end_year {
                    Some(_) => closed_timespans += 1,
                    None => open_timespans += 1,
                },
            }
        }
        for artist in artist.related.0 {
            let id = artist.id;
            if !known_artist_ids.contains(&id) {
                new_artist_ids.push_back(artist.id);
            }
        }
        print!(
            "\rknown good artists: {} (decades: {decades}, \
            closed: {closed_timespans}, open: {open_timespans})",
            known_artist_ids.len()
        );
        std::io::stdout().flush().unwrap();
    }
}

```

</details>

### CODE REVIEW

File: metadata/src/artist.rs [hunk 1]
```diff
@@ -111,10 +111,12 @@ pub struct Biographies(pub Vec<Biography>);
...
-pub struct ActivityPeriod {
-    pub start_year: i32,
-    pub end_year: i32,
-    pub decade: i32,
+pub enum ActivityPeriod {
+    Timespan {
+        start_year: i32,
+        end_year: Option<i32>,
+    },
+    Decade(i32),
```
Comment (eladyn) [inline metadata/src/artist.rs:119]: Would it make sense to make these three numbers unsigned (e.g. `u16` or even less) to potentially save space?

**Comment by roderickvd** [comment]
Do you consider this purely as an utility, or would you want to use this as a type in the metadata structs? As I see that you introduce it but do not add usage of it anywhere.

> I also added two `debug_assert!`s that make sure that the invariants that I assume are always held up. Alternatively, I could also turn them into normal `assert!`s or handle-able exceptions through `TryFrom`. I went for the current variants, since I think it makes most sense, but I'm open to different perspectives!

If it's fallible which it does seem like (after all, there is no guarantee that any of the fields are set) then it should be `TryFrom` I think.

File: metadata/src/artist.rs (hunk 1 above)
Comment (roderickvd) [reply metadata/src/artist.rs:119]: I'm pretty sure we could. Probably Spotify sends everything as `i32` because there are no 16-bit integers in protobuf?

**Comment by eladyn** [comment]
> Do you consider this purely as an utility, or would you want to use this as a type in the metadata structs? As I see that you introduce it but do not add usage of it anywhere.

Well, it is already used in `Artist` if that is what you meant. Although you're right that no librespot code actually accesses this field, I just came across it while playing with the different metadata endpoints and thought that it might be (small), but potentially usable change.

> If it's fallible which it does seem like (after all, there is no guarantee that any of the fields are set) then it should be `TryFrom` I think.

Thinking about it a bit more, that makes sense; I will change it.

File: metadata/src/artist.rs (hunk 1 above)
Comment (eladyn) [reply metadata/src/artist.rs:119]: Apparently, there would at least be [unsigned integers](https://developers.google.com/protocol-buffers/docs/proto#scalar). I'll leave the decision up to you: Do we want to not fail if Spotify sends us strange (negative) values? If we decide to go for conversion to `u16`, should this be fallible or is it OK to use `as u16` there (which would of course panic for negative values).

Comment (roderickvd) [reply metadata/src/artist.rs:119]: Let’s return an error, certainly not panic in a library.

**Comment by roderickvd** [comment]
OK great!

**Comment by eladyn** [comment]
Everything should be fine now. Let me know what you think of this version. :)

**Comment by roderickvd** [comment]
LGTM, smart!

**Pull request merged**